### PR TITLE
Post Processor for ColumnNameModel

### PR DESCRIPTION
**Pull request by taylorfturner**

### CODE REVIEW

**Comment by taylorfturner** [comment]
#610

File: dataprofiler/tests/labelers/test_data_processing.py [hunk 1]
```diff
@@ -2928,3 +2930,69 @@ def test_random_state_constructor(self):
...
+                    "attirbute": "test",
+                    "label": "test",
+                }
+            ],
+            "positive_threshold_config": 85,
+        }
+        post_processor = ColumnNameModelPostprocessor(parameters=params)
+
+        # call save processor func
+        post_processor._save_processor(".")
```
Comment (taylorfturner) [inline dataprofiler/tests/labelers/test_data_processing.py:2998]: I despise this test... really need it mocked

File: dataprofiler/labelers/column_name_model.py [hunk 2]
```diff
@@ -36,6 +36,8 @@ def __init__(self, parameters=None):
             parameters = {}
         parameters.setdefault("false_positive_dict", None)
         parameters.setdefault("true_positive_dict", None)
+        parameters.setdefault("include_label", True)
+        parameters.setdefault("negative_threshold_config", None)
```
Comment (taylorfturner) [inline dataprofiler/labelers/column_name_model.py:40]: params for the `column_name_model` and not `predict()` parameters

File: dataprofiler/tests/labelers/test_column_name_model.py [hunk 3]
```diff
@@ -113,51 +94,22 @@ def test_help(self, mock_stdout):
...
-        model_output = model.predict(
-            data=["ssn", "role_name", "wallet_address"], show_confidences=True
-        )
-        self.assertTrue(np.array_equal(expected_output, model_output))
+        # `show_confidences` is disabled currently
+        # should raise error if set to `True`
+        with self.assertRaises(NotImplementedError):
+            model.predict(
+                data=["ssn", "role_name", "wallet_address"], show_confidences=True
+            )
```
Comment (taylorfturner) [inline dataprofiler/tests/labelers/test_column_name_model.py:112]: disabled `show_confidences` right now. Need to clean this up and enable it as a toggle

File: dataprofiler/labelers/column_name_model.py [hunk 4]
```diff
@@ -112,35 +136,6 @@ def _compare_negative(
...
-            if scores[i][0] > positive_threshold:
-                output_dictionary[list_of_column_names[i]] = {}
-                output_dictionary[list_of_column_names[i]]["pred"] = check_values_dict[
-                    scores[i][1]
-                ]["label"]
-                if show_confidences:
-                    output_dictionary[list_of_column_names[i]]["conf"] = scores[i][0]
-
-        return output_dictionary
-
```
Comment (taylorfturner) [inline dataprofiler/labelers/column_name_model.py:143]: completely got rid of the `_compare_positive` becuase we don't need it anymore after moving some things to the post processor

File: dataprofiler/labelers/column_name_model.py [hunk 5]
```diff
@@ -186,9 +181,6 @@ def predict(
         batch_size=None,
         show_confidences=False,
         verbose=True,
-        include_label=True,
-        negative_threshold_config=50,
-        positive_threshold_config=85,
```
Comment (taylorfturner) [inline dataprofiler/labelers/column_name_model.py:191]: parameters to the class __init__ not the predict method. Keeping the API consistent across all the models

File: dataprofiler/labelers/column_name_model.py [hunk 6]
```diff
@@ -208,18 +201,29 @@ def predict(
...
+            scorer=rapidfuzz.fuzz.token_sort_ratio,
+            include_label=self._parameters["include_label"],
         )
+
+        if show_confidences:
+            raise NotImplementedError(
+                """`show_confidences` parameter is disabled
+                for Proof of Concept implementation. Confidence
+                values are enabled by default."""
+            )
```
Comment (taylorfturner) [inline dataprofiler/labelers/column_name_model.py:225]: Need to think through how to enable this

File: dataprofiler/labelers/data_processing.py [hunk 7]
```diff
@@ -2079,3 +2079,129 @@ def process(self, data, labels=None, label_mapping=None, batch_size=None):
...
+                the order in which entities are given priority must be set
+            random_state: Random state setting to be used for randomly
+                selecting a prediction when two labels have equal opportunity
+                for a given sample.
+        :type parameters: dict
+        :return: None
+        """
+        # allowed_parameters = self.__class__.__init__.__code__.co_varnames[
+        #     1 : self.__class__.__init__.__code__.co_argcount
+        # ]
```
Comment (taylorfturner) [inline dataprofiler/labelers/data_processing.py:2116]: don't think we need this anymore for this model @JGSweets

File: dataprofiler/labelers/data_processing.py [hunk 8]
```diff
@@ -2079,3 +2079,129 @@ def process(self, data, labels=None, label_mapping=None, batch_size=None):
...
+            cls.__name__
+            + "\n\n"
+            + "Parameters:\n"
+            + param_docs[param_start_ind:param_end_ind]
+            + "\nProcess Output Format:\n"
+            "    Each sample receives a label.\n"
+            "    Original data - ['My', 'String', ...]\n"
+            "    Output labels - ['<LABEL_1>', '<LABEL_2>', "
+            "..(num samples)]"
+        )
```
Comment (taylorfturner) [inline dataprofiler/labelers/data_processing.py:2176]: needs an update

File: dataprofiler/tests/labelers/test_data_processing.py [hunk 9]
```diff
@@ -2928,3 +2930,69 @@ def test_random_state_constructor(self):
...
+            ],
+            "positive_threshold_config": 85,
+        }
+
+        processor = ColumnNameModelPostprocessor(parameters=params)
+        process_output = processor.process(data)
+
+        self.assertEqual(process_output, expected_output)
+
+    def test_save_processor(self):
```
Comment (JGSweets) [inline dataprofiler/tests/labelers/test_data_processing.py:2985]: so this would need to be mocked bc otherwise it saves the file to disk

Comment (JGSweets) [reply dataprofiler/tests/labelers/test_data_processing.py:2985]: `@mock.patch("builtins.open")`

Comment (JGSweets) [reply dataprofiler/tests/labelers/test_data_processing.py:2985]: if we use the other examples, can then use:
`mock_file = setup_save_mock_open(mock_open)`

Comment (JGSweets) [reply dataprofiler/tests/labelers/test_data_processing.py:2985]: Check out https://github.com/capitalone/DataProfiler/pull/629/files#diff-fd672afb661ad4d53cbd966058bbb9f993407275e507567950ae75290e45cc8eR2896
for the rest, just above in regex's post_save

File: dataprofiler/labelers/column_name_model.py (hunk 2 above)
Comment (taylorfturner) [reply dataprofiler/labelers/column_name_model.py:40]: keeping the `predict()` API consistent

File: dataprofiler/labelers/column_name_model.py [hunk 10]
```diff
@@ -77,14 +87,28 @@ def _validate_parameters(self, parameters):
                 not isinstance(value, list)
                 or not isinstance(value[0], dict)
                 and "attribute" not in value[0].keys()
-                and "list" not in value[0].keys()
+                and "label" not in value[0].keys()
```
Comment (taylorfturner) [inline dataprofiler/labelers/column_name_model.py:90]: typo

File: dataprofiler/tests/labelers/test_column_name_model.py (hunk 3 above)
Comment (taylorfturner) [reply dataprofiler/tests/labelers/test_column_name_model.py:112]: an option is to put this in the post processor part of the labeler, but that breaks the API paradigm for the `model's` `predict()` method ...

File: dataprofiler/labelers/data_processing.py [hunk 11]
```diff
@@ -2079,3 +2079,113 @@ def process(self, data, labels=None, label_mapping=None, batch_size=None):
...
+):
+    """Subclass of BaseDataPostprocessor for postprocessing regex data."""
+
+    def __init__(self, parameters):
+        """Initialize the ColumnNameModelPostProcessor class."""
+        # set parameters
+        if not parameters:
+            parameters = {}
+        parameters.setdefault("true_positive_dict", None)
+        parameters.setdefault("positive_threshold_config", None)
```
Comment (JGSweets) [inline dataprofiler/labelers/data_processing.py:2095]: this ultimately should mimic the other procesors where we take in paramters as individual args: e.g.

```python
def __init__(
        self,
        default_label="UNKNOWN",
        pad_label="PAD",
        flatten_separator=" ",
        use_word_level_argmax=False,
        output_format="character_argmax",
        separators=(" ", ",", ";", "'", '"', ":", "\n", "\t", "."),
        word_level_min_percent=0.75,
    ):
        """
        Initialize the CharPostprocessor class.

        :param default_label: Key for label_mapping that is the default label
        :type default_label: string (could be int, char, etc.)
        :param pad_label: Key for label_mapping that is the pad label
        :type pad_label: string (could be int, char, etc.)
        :param flatten_separator: separator used to put between flattened
            samples.
        :type flatten_separator: str
        :param use_word_level_argmax: whether to require the argmax value of
            each character in a word to determine the word's entity
        :type use_word_level_argmax: bool
        :param output_format: (character_argmax vs NER) where character_argmax
            is a list of encodings for each character in the input text and NER
            is in the dict format which specifies start,end,label for each
            entity in a sentence
        :type output_format: str
        :param separators: list of characters to use for separating words within
            the character predictions
        :type separators: tuple(str)
        :param word_level_min_percent: threshold on generating dominant
            word_level labeling
        :type word_level_min_percent: float
        """
        super().__init__(
            default_label=default_label,
            pad_label=pad_label,
            flatten_separator=flatten_separator,
            use_word_level_argmax=use_word_level_argmax,
            output_format=output_format,
            separators=separators,
            word_level_min_percent=word_level_min_percent,
        )
```

Comment (JGSweets) [reply dataprofiler/labelers/data_processing.py:2095]: the super should auto handle parameters where need be,
if we want `true_positive_dict` to be a dict by default, we could always do a 
```python
if true_positive_dict is None:
    true_positive_dict = {}
```

File: dataprofiler/labelers/data_processing.py [hunk 12]
```diff
@@ -2079,3 +2079,113 @@ def process(self, data, labels=None, label_mapping=None, batch_size=None):
...
+    BaseDataPostprocessor, metaclass=AutoSubRegistrationMeta
+):
+    """Subclass of BaseDataPostprocessor for postprocessing regex data."""
+
+    def __init__(self, parameters):
+        """Initialize the ColumnNameModelPostProcessor class."""
+        # set parameters
+        if not parameters:
+            parameters = {}
+        parameters.setdefault("true_positive_dict", None)
```
Comment (JGSweets) [inline dataprofiler/labelers/data_processing.py:2094]: do we want these defaults to both be `None`, will that error by default or do we want to use any defaults?

File: dataprofiler/labelers/data_processing.py [hunk 13]
```diff
@@ -2079,3 +2079,113 @@ def process(self, data, labels=None, label_mapping=None, batch_size=None):
...
+        parameters.setdefault("positive_threshold_config", None)
+
+        super().__init__(**parameters)
+
+    def _validate_parameters(self, parameters):
```
Comment (JGSweets) [inline dataprofiler/labelers/data_processing.py:2099]: I think we are missing tests for this

**Review by JGSweets — CHANGES_REQUESTED** [review]

**Review by micdavis — APPROVED** [review]

File: dataprofiler/labelers/data_processing.py (hunk 13 above)
Comment (taylorfturner) [reply dataprofiler/labelers/data_processing.py:2099]: added

File: dataprofiler/labelers/data_processing.py (hunk 12 above)
Comment (taylorfturner) [reply dataprofiler/labelers/data_processing.py:2094]: - `parameters.setdefault("true_positive_dict", None)` default for this is a no go in my opinion unless we want "out of the box" examples of attributes and labels, which I'd say is out of scope for this PR
- `parameters.setdefault("positive_threshold_config", None)`: however a default for this is a fair point, but really only matters if the `true_positive_dict` is set as well...

File: dataprofiler/labelers/data_processing.py (hunk 11 above)
Comment (taylorfturner) [reply dataprofiler/labelers/data_processing.py:2095]: ah good point... I was following the `model` paradigm still. I'll update this

**Review by micdavis — DISMISSED** [review]

**Review by micdavis — DISMISSED** [review]

**Review by JGSweets — APPROVED** [review]

**Review by micdavis — APPROVED** [review]

**Pull request merged**